src/wix/WixToolset.Converters/WixConverter.cs

```diff
@@ -2854,6 +2854,45 @@ private static void RenameElementToStandardDirectory(XElement element)
             }
         }
 
+        // If TryGetInnerText succeeds, normalize the string value it returns by replacing all \r's \n's with spaces,
+        // then replacing all multiple spacws with single spaces. This is suitable for use in an XText representing a condition.
```

Comment:
I think this is dangerous as it could replace double spaces in a quoted value with a single space, breaking the original intent. For example `PROP == "Foo  Bar"` where there are two spaces between Foo and Bar would get modified.

If anything, you could split the XML on `\r` and `\n`, removing empty lines then `.Trim()` the resulting values to cut the leading/trailing whitespace away then `.Join(' ')` the result back together.

But looping through each character required understanding the context for each use of inner text.

Reply:
I hadn't considered whitespace between quotes. Is non-blank whitespace explicitly disallowed by the grammar? Doe it, or should it produce an error? Let me give an example, using \r as carriage return, \n as new line, and \b as space.

<Condition>
\b\bvariable = "abc\b\b\r\nxyz"
<Condition/>

This would split the condition into

\b\bvariable = "abc\b\b


xyz"

Removing the empty lines then yields:

\b\bvariable = "abc\b\b
xyz"

Then, trimming both lines yields:

variable = "abc
xyz"

and joining with \n yields:

variable = "abc\nxyz"

This puts us back where we began without considering quotes, i.e. the XText constructor will create a value for Condition of "abc\u000axyz", which certainly won't do anything like what you might have had in mind.

I'm hoping that the non-space white space already generates an error when found in quotes, and that I will never see this construct. If it doesn't, then I'd like to produce an error (of what severity and through what messaging mechanism. Finally, I'd want to delete the offending characters and let the chips fall where they might.

I'm assuming that both species of quotes would be seen literally in the incoming inner text, Once stored as an attribute value, will the quotes need to be escaped or all converted to single quotes so that they can be used in an

attribute="value"

setting

Reply:
I have rewritten TryGetNormalizedInnerText(). Highlights are:

- Replace contiguous unquoted whitespace with a single space.
- Use single quotes for all outgoing outer quotes.
- Escape outgoing nested quotes with backslash.
- Escape all incoming backslashes with backslash.
- Trim the resulting string.

I hope I haven't overlooked anything.